karapace/auth.py

```diff
@@ -0,0 +1,185 @@
+from base64 import b64encode
+from dataclasses import dataclass, field
+from enum import Enum, unique
+from hmac import compare_digest
+from karapace.config import InvalidConfiguration
+from karapace.rapu import JSON_CONTENT_TYPE
+from typing import Optional
+
+import aiohttp
+import aiohttp.web
+import argparse
+import asyncio
+import base64
+import hashlib
+import json
+import logging
+import os
+import re
+import secrets
+import sys
+
+log = logging.getLogger(__name__)
+
+
+@unique
+class Operation(Enum):
+    Read = "Read"
+    Write = "Write"
+
+
+def hash_password(algorithm: str, salt: str, password: str) -> str:
+    if algorithm in ["sha1", "sha256", "sha512"]:
+        return b64encode(
+            hashlib.pbkdf2_hmac(algorithm, bytearray(password, "UTF-8"), bytearray(salt, "UTF-8"), 5000)
+        ).decode("ascii")
+    if algorithm == "scrypt":
+        return str(
+            base64.b64encode(hashlib.scrypt(bytearray(password, "utf-8"), salt=bytearray(salt, "utf-8"), n=16384, r=8, p=1)),
+            encoding="utf-8",
+        )
+    raise NotImplementedError(f"Hash algorithm '{algorithm}' is not implemented")
+
+
+@dataclass
+class User:
+    username: str
+    algorithm: str
+    salt: str
+    password: str = field(repr=False)  # hashed
+
+    def compare_password(self, plaintext: str) -> bool:
+        return compare_digest(self.password, hash_password(self.algorithm, self.salt, plaintext))
+
+
+@dataclass(frozen=True)
+class ACL:
```

Comment:
Nit: this in an access control list entry, not an ACL. Also, later in the code class is used in an array called `acls`. That is not not acls, as in access control lists, but a list of acl entries. Thus this would be more correct:

```python
class ACLEntry:
    ....

acl = [ACLEntry(x, y, z), ACLEntry(a, b, c)
```